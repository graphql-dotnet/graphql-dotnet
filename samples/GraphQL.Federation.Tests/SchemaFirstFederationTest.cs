using System.Text.Json;
using GraphQL.DataLoader;
using GraphQL.Federation.Tests.Fixtures;
using GraphQL.SystemTextJson;

namespace GraphQL.Federation.Tests;

// dotnet test GraphQL.Federation.Tests --filter DisplayName~GraphQL.Federation.Tests.SchemaFirstFederationTest.ServiceTest
// dotnet test GraphQL.Federation.Tests --filter DisplayName~GraphQL.Federation.Tests.SchemaFirstFederationTest.EntitiesTest
[Collection(nameof(SchemaFirstCollectionDefinition))]
public class SchemaFirstFederationTest : BaseSchemaFirstGraphQLTest
{
    public SchemaFirstFederationTest(SchemaFirstFixture schemaFirstFixture)
        : base(schemaFirstFixture)
    { }

    [Fact]
    public async Task ServiceTest()
    {
        // WaitForDebugger();

        var json = await Schema.ExecuteAsync(new GraphQLSerializer(), options =>
        {
            options.Listeners.Add(new DataLoaderDocumentListener(new DataLoaderContextAccessor()));
            options.ThrowOnUnhandledException = true;
            options.Query = """
                query {
                    _service {
                        sdl
                    }
                }
                """;
        });

        var sdl = JsonDocument.Parse(json).RootElement
            .GetProperty("data")
            .GetProperty("_service")
            .GetProperty("sdl")
            .GetString()!;

        sdl.ShouldMatchApproved(c => c.NoDiff());
    }

    [Fact]
    public async Task EntitiesTest()
    {
        // WaitForDebugger();

        var json = await Schema.ExecuteAsync(new GraphQLSerializer(), options =>
        {
            options.RequestServices = Services;
            options.Listeners.Add(new DataLoaderDocumentListener(new DataLoaderContextAccessor()));
            options.ThrowOnUnhandledException = true;
            options.Query = """
                query {
                    _entities(representations: [{ __typename: "SchemaFirstFederatedTestDto", id: 1 }, { __typename: "SchemaFirstFederatedTestDto", id: 3 }, { __typename: "SchemaFirstExternalResolvableTestDto", id: 321, external: "zxcvbn" }]) {
                        ... on SchemaFirstFederatedTestDto {
                            name
                            externalTest {
                                id
                            }
                            externalResolvableTest {
                                id
                                external
                                extended
                            }
                        }
                        ... on SchemaFirstExternalResolvableTestDto {
                            id
                            extended
                        }
                    }
                }
                """;
        });

        json.ShouldBe(
            """{"data":{"_entities":[{"name":"111","externalTest":{"id":4},"externalResolvableTest":{"id":7,"external":"qwerty","extended":"qwerty"}},{"name":"333","externalTest":{"id":6},"externalResolvableTest":{"id":9,"external":"qwerty","extended":"qwerty"}},{"id":321,"extended":"zxcvbn"}]}}""",
            StringCompareShould.IgnoreLineEndings);
    }
}
