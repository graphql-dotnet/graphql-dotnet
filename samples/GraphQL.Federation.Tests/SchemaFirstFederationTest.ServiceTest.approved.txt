schema @link(url: "https://specs.apollo.dev/link/v1.0", import: ["@link"]) @link(url: "https://specs.apollo.dev/federation/v2.3", import: ["@key", "@external", "@requires", "@provides", "@shareable", "@inaccessible", "@override", "@tag"]) {
  query: Query
}

scalar _Any

type Query {
  _noop: String
  _service: _Service!
  _entities(representations: [_Any!]!): [_Entity]!
}

type SchemaFirstExternalResolvableTestDto @key(fields: "id") {
  id: Int!
  external: String @external
  extended: String! @requires(fields: "external")
}

type SchemaFirstExternalTestDto @key(fields: "id", resolvable: false) {
  id: Int!
}

type SchemaFirstFederatedTestDto @key(fields: "id") {
  id: Int!
  name: String @deprecated(reason: "Test deprecation reason 03.")
  externalTestId: Int!
  externalResolvableTestId: Int!
  externalTest: SchemaFirstExternalTestDto! @deprecated(reason: "Test deprecation reason 04.")
  externalResolvableTest: SchemaFirstExternalResolvableTestDto! @provides(fields: "external")
}

type _Service {
  sdl: String
}

union _Entity = SchemaFirstExternalResolvableTestDto | SchemaFirstFederatedTestDto
